### DISCUSSION TITLE
Non-blocking Input Validation Workaround

### DISCUSSION THREAD

**Discussion by KaiAragaki** [charmbracelet/bubbles#306] [Q&A] [answered]
Hi all!

# Related
This is probably related to #289 and #185 

# Goal

To produce a field that looks like this:

```
> 
```
And when the input is added...

```
> my favorite name is Larry    | x Capital letters are not allowed
```
...a warning will appear on the side denoting incorrect input.

# What I've tried

I've tried providing a [`ValidateFunc`](https://pkg.go.dev/github.com/charmbracelet/bubbles/textinput#ValidateFunc) to the input, but it is *blocking*. That is, input stops at

```
> my favorite name is 
```
and will only allow input that follows the rules set by `ValidateFunc`.

# What I'm requesting

I understand that #185 seeks to get around this by modifying the current API. However, for me, right now, this is project blocking, so I wonder if there may be a temporary work around. I've tried my best, but unfortunately I'm learning Go and the charm-verse at the same time, and am a bit over my head!

**Comment by bashbunni** [comment]
Sorry for the delay! I'll take a look 👀

**Comment by bashbunni** [comment, marked as answer]
I see what you mean about how `textinput.Validate` blocks the behaviour. In this case, you could just define your own validation function that gets called in `Update` after the user types and assign the error to the model's error field, then check that in `View` and render accordingly. I've attached a Gist with my solution that gives the outcome shown in the attached image. Of course you can tweak styling, etc as needed. Hope that helps! Let us know if you still have questions

![image](https://user-images.githubusercontent.com/15822994/211713969-74e0d2a3-d294-4e26-a465-b1db17b1da34.png)
https://gist.github.com/bashbunni/4573b0fc981123875a835f1a0aa412ca

PS: If you want the formatting to match your example: `> my favorite name is Larry    | x Capital letters are not allowed` you can use `lipgloss.JoinHorizontal` instead of `lipgloss.JoinVertical`
